documentation/design-docs/dotnet-tools.md

```diff
@@ -1,124 +1,801 @@
+
+    -o, --output
+        The path where the converted file is written. If unspecified the file is written in the current directory
+        using the same base filename as the input file and the extension appropriate for the new format.
+
+    trace_file_path
+        The path to the trace file that should be converted. The trace file can be in either a netperf or netperf.zip file.
+
+    Examples:
+      > dotnet-trace convert --format speedscope trace.netperf
+      Writing:       ./trace.speedscope.json
+      Conversion complete
+
+PACK
+
+    dotnet-trace pack [-h|--help]
+                      [-o|--output <output_file_path>]
+                      [--verbose]
+                      <trace_file_path>
+
+    Compresses a trace and any necessary symbols into a single zip file for easy off-machine analysis     
+
+    -h, --help
+        Show command line help
+
+    -o, --output
+        The path where the pack is written. If unspecified the pack is written in the current directory
+        using the same base filename as the input file and the .zip extension.
+
+    --verbose
+        Logs detailed information about what the pack command is doing.
+
+    trace_file_path
+        The path to the trace file that should be packed.
+
+
+    Examples:
+      > dotnet-trace pack trace.netperf
+      Packing:      ./trace.netperf.zip
+      Pack complete
+
+      > dotnet-trace pack --verbose trace.netperf
+      Packing:      /usr/home/noahfalk/trace.netperf.zip
+      Compressing   /usr/home/noahfalk/trace.netperf
+      Checking      /usr/bin/dotnet/shared/3.0.170/System.Private.CoreLib.dll
+        Not packing symbols - Policy skips Microsoft binary
+      Checking      /usr/bin/dotnet/shared/3.0.170/System.Diagnostics.dll
+        Not packing symbols - Policy skips Microsoft binary
+      Checking      /usr/home/noahfalk/MyApp/Newtonsoft.Json.dll
+        Searching for Newtonsoft.Json.pdb
+        Searching   /usr/home/noahfalk/MyApp/Newtonsoft.Json.pdb
+        Not packing symbols - Newtonsoft.Json.pdb not found
+      Checking      /usr/home/noahfalk/MyApp/MyApp.dll
+        Searching for MyApp.pdb
+        Searching   /usr/home/noahfalk/MyApp/MyApp.pdb
+        Found matching symbol file
+        Compressing symbol file /usr/home/noahfalk/MyApp/MyApp.pdb
+      ...
+      Pack Complete
+
+### dotnet-dump ###
 
 SYNOPSIS
 
-    dotnet collect dump [-h, --help]
-                        [--provider]
-                        [--buffer]
-                        
+    dotnet-dump [--version] 
+                [-h, --help]
+                <command> [<args>]
 
 OPTIONS
 
+    --version
+        Display the version of the dotnet-dump utility.
+
     -h, --help
-        Prints the synopsis and a list of the most commonly used commands.
+        Show command line help
+
+COMMANDS
+
+    collect   Capture one or more dumps (core files on Mac/Linux) from a process
+    analyze   Starts an interactive shell with debugging commands to explore a dump
```

Comment:
Can you use commands like `GCHEAPDIFF` from the commandline (as opposed to an interactive shell) if you want to script generation and analysis of dumps?

Reply:
In my current proposal no, but I'm not at all opposed to opening that up. I could see how scripting heapdiff specifically wouldn't be too hard to do and might be useful. As we go into other commands the customer value might decrease for a couple reasons:
a) A debugger-like REPL (which this is, albeit a very limited one) typically has some state such as current thread or current frame. That state could be specified on the shell command line explicitly but it makes each command more verbose.
b) Many commands make use of data that was discovered by previous commands, such as DumpHeap printing object addresses and then GCRoot exploring one of them. For a script to be useful it would need to do text processing on previous commands, then implement logic that replicates the human's choice about what is interesting, then pass that argument to the next command. Typically when you get to this level of complexity you wish that you had a strongly typed API to program against rather than a slow/awkward/version-brittle text scraping interface.